### PR TITLE
Fix issues in loading GMT's shared library

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

The `load_libgmt()` function still has some bugs:

https://github.com/GenericMappingTools/pygmt/blob/6a6171e3efc112f2a8a86088e24caf932149e928/pygmt/clib/loading.py#L36-L52

1. At line 45, we only catch the `OSError` exception, but `check_libgmt()` may raise `GMTCLibError` exception. So if the library paths are `["/path/to/GMT5/libgmt.so", "/path/to/GMT6/libgmt.so", "libgmt.so"]`, the `load_libgmt()` function won't be able to the working library.
2. After #702, the `clib_full_names()` function returns a list of all possible paths. Some paths may be duplicated. So, if the paths are `["invalid_path_A", "invalid_path_A", "invalid_path_B"]`, we will check `invalid_path_A` twice. 
3. The variable `error` only contains the error message of the last library path, but, each library path may fail due to different reasons.

This PR makes the following changes:

1. Catch the `GMTCLibError` error from calling `check_libgmt()` (Fixes point 1)
2. Skip a library path if it's known to fail in previous tries (Fixes point 2)
3. Improve the error message, by combine error message of all tries (Fixes point 3)
4. Add a new parameter `lib_fullnames` (default to `clib_full_names()`) to for easier testing
4. Add more tests to make sure it works.

Fixes #

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:117]:
```suggestion
        elif isinstance(libname, str):
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:173]: To be honest, I kind of wanted you to move the tests for clib_full_names to a separate file during #872 (e.g. to 'test_clib_loading_full_names'?), wondering if we should do it here since this test file is getting really complicated.

**Review by weiji14 — CHANGES_REQUESTED** [review]
Quite a few comments. After trying to have a go at refactoring the big 'test_load_libgmt_with_broken_libraries', I can see why you've coded it up this way, but it's not easy to follow the code and I worry about the poor maintainer in the future having to understand what is happening :cry:

In a way, I wonder if we're overcomplicating the `load_libgmt` et al. functions in `pygmt/clib/loading.py`. It seems like a crazy amount of possible combinations we're testing (3 OSes, different GMT_LIBRARY_PATHs, libraries that may or may not load, etc). Probably need to have a hard think about whether we can actually simplify the whole loading logic.

File: pygmt/tests/test_clib_loading.py
Comment (seisman) [reply pygmt/tests/test_clib_loading.py:117]: pylint doesn't like this change.

File: pygmt/clib/loading.py
Comment (seisman) [reply pygmt/clib/loading.py:44]: > No need for an `error` variable, can just check if `error_msg` is empty or not.

No, we can't check `error_msg`. Assume that we have two paths to try, the first path fails but the second path succeeds, then error_msg will contain the messages for the first path, but `error` is still `False`.

File: pygmt/clib/loading.py
Comment (seisman) [reply pygmt/clib/loading.py:42]: It's shorter, but the old code is easier to understand, no?

File: pygmt/tests/test_clib_loading.py
Comment (seisman) [reply pygmt/tests/test_clib_loading.py:173]: I'm OK with moving the tests into separate files.

File: pygmt/clib/loading.py
Comment (weiji14) [reply pygmt/clib/loading.py:42]: Ok, fair enough.

File: pygmt/clib/loading.py
Comment (weiji14) [reply pygmt/clib/loading.py:44]: Ah yes, forgot about that. We could just keep the `error` variable, or clear `error_msg` (i.e. set `error_msg=[]`) when a library is successfully loaded perhaps?

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [reply pygmt/tests/test_clib_loading.py:117]: Oh ok, must be the return statement above, Ignore this then.

File: pygmt/clib/loading.py
Comment (seisman) [reply pygmt/clib/loading.py:44]: Yes, either should work, but I prefer to keep the `error` variable to make the code more readable.

File: pygmt/tests/test_clib_loading.py
Comment (seisman) [reply pygmt/tests/test_clib_loading.py:173]: Perhaps still keep the tests in the same file, because it's only ~300 lines, so still maintainable.

I've used comment lines to group the tests:
```
###############################################################################
# Tests for clib_full_names
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:92]:
```suggestion
class TestLibgmtBrokenLibs:
```

Turn this big test into a class with multiple sub-tests (see https://docs.pytest.org/en/4.6.x/getting-started.html#group-multiple-tests-in-a-class). A lot more lines of code, but the advantage is that we can easily see which test broke exactly in the pytest log (if things break in the future).

```python-traceback
pygmt/tests/test_clib_loading.py::TestLibgmtBrokenLibs::test_two_broken_libraries PASSED [ 37%]
pygmt/tests/test_clib_loading.py::TestLibgmtBrokenLibs::test_load_brokenlib_invalidpath PASSED [ 43%]
pygmt/tests/test_clib_loading.py::TestLibgmtBrokenLibs::test_brokenlib_invalidpath_workinglib PASSED [ 50%]
pygmt/tests/test_clib_loading.py::TestLibgmtBrokenLibs::test_invalidpath_brokenlib_workinglib PASSED [ 56%]
pygmt/tests/test_clib_loading.py::TestLibgmtBrokenLibs::test_workinglib_brokenlib_invalidpath PASSED [ 62%]
pygmt/tests/test_clib_loading.py::TestLibgmtBrokenLibs::test_brokenlib_brokenlib_workinglib PASSED [ 68%]
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:94]:
```suggestion
    Test that load_libgmt still works when a broken library is found.
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:99]:
```suggestion

    # load the GMT library before mocking the ctypes.CDLL function
    loaded_libgmt = load_libgmt()
    invalid_path = "/invalid/path/to/libgmt.so"
    faked_libgmt1 = FakedLibGMT("/path/to/faked/libgmt1.so")
    faked_libgmt2 = FakedLibGMT("/path/to/faked/libgmt2.so")

    def _mock_ctypes_cdll_return(self, libname):
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:130]:
```suggestion
        return self.loaded_libgmt

    @pytest.fixture
    def mock_ctypes(self, monkeypatch):
        monkeypatch.setattr(ctypes, "CDLL", self._mock_ctypes_cdll_return)
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:140]:
```suggestion
    def test_two_broken_libraries(self, mock_ctypes):
        """
        Case 1: two broken libraries Raise the GMTCLibNotFoundError exception.

        Error message should contain information of both libraries that failed
        to load properly.
        """
        lib_fullnames = [self.faked_libgmt1, self.faked_libgmt2]
        msg_regex = (
            fr"Error loading GMT shared library at '{self.faked_libgmt1._name}'.\n"
            fr"Error loading '{self.faked_libgmt1._name}'. Couldn't access.*\n"
            fr"Error loading GMT shared library at '{self.faked_libgmt2._name}'.\n"
            f"Error loading '{self.faked_libgmt2._name}'. Couldn't access.*"
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:151]:
```suggestion
    def test_load_brokenlib_invalidpath(self, mock_ctypes):
        """
        Case 2: broken library + invalid path Raise the GMTCLibNotFoundError
        exception.

        Error message should contain information of one library that failed to
        load and one invalid path.
        """
        lib_fullnames = [self.faked_libgmt1, self.invalid_path]
        msg_regex = (
            fr"Error loading GMT shared library at '{self.faked_libgmt1._name}'.\n"
            fr"Error loading '{self.faked_libgmt1._name}'. Couldn't access.*\n"
            fr"Error loading GMT shared library at '{self.invalid_path}'.\n"
            f"Unable to find '{self.invalid_path}'"
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:157]:
```suggestion
    def test_brokenlib_invalidpath_workinglib(self, mock_ctypes):
        """
        Case 3: broken library + invalid path + working library.
        """
        lib_fullnames = [self.faked_libgmt1, self.invalid_path, self.loaded_libgmt]
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:161]:
```suggestion
    def test_invalidpath_brokenlib_workinglib(self, mock_ctypes):
        """
        Case 4: invalid path + broken library + working library.
        """
        lib_fullnames = [self.invalid_path, self.faked_libgmt1, self.loaded_libgmt]
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:165]:
```suggestion
    def test_workinglib_brokenlib_invalidpath(self, mock_ctypes):
        """
        Case 5: working library + broken library + invalid path.
        """
        lib_fullnames = [self.loaded_libgmt, self.faked_libgmt1, self.invalid_path]
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:169]:
```suggestion
    def test_brokenlib_brokenlib_workinglib(self, mock_ctypes):
        """
        Case 6: repeating broken libraries + working library.
        """
        lib_fullnames = [self.faked_libgmt1, self.faked_libgmt1, self.loaded_libgmt]
```

**Review by weiji14 — APPROVED** [review]
Thanks for adding the comprehensive tests @seisman, sorry for dragging this out for so long. Will be good to merge once all the tests pass and the Style Checks errors are resolved.

**Comment by seisman** [comment]
@weiji14 Thanks for your review.

Now the Style Checks reports errors like:
```
pygmt/tests/test_clib_loading.py:135:40: W0613: Unused argument 'mock_ctypes' (unused-argument)
```
The argument `mock_ctypes` is necessary, so I can't remove it. Should I just disable the pylint warning?

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:135]:
```suggestion
    def test_two_broken_libraries(self, mock_ctypes):  # pylint: disable=unused-argument
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:153]:
```suggestion
    def test_load_brokenlib_invalidpath(
        self, mock_ctypes
    ):  # pylint: disable=unused-argument
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:171]:
```suggestion
    def test_brokenlib_invalidpath_workinglib(
        self, mock_ctypes
    ):  # pylint: disable=unused-argument
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:178]:
```suggestion
    def test_invalidpath_brokenlib_workinglib(
        self, mock_ctypes
    ):  # pylint: disable=unused-argument
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:185]:
```suggestion
    def test_workinglib_brokenlib_invalidpath(
        self, mock_ctypes
    ):  # pylint: disable=unused-argument
```

File: pygmt/tests/test_clib_loading.py
Comment (weiji14) [inline pygmt/tests/test_clib_loading.py:192]:
```suggestion
    def test_workinglib_brokenlib_invalidpath(
        self, mock_ctypes
    ):  # pylint: disable=unused-argument
```

**Review by weiji14 — COMMENTED** [review]
> Now the Style Checks reports errors like:
> 
> ```
> pygmt/tests/test_clib_loading.py:135:40: W0613: Unused argument 'mock_ctypes' (unused-argument)
> ```
> 
> The argument `mock_ctypes` is necessary, so I can't remove it. Should I just disable the pylint warning?

Yes.

**Pull request merged**
